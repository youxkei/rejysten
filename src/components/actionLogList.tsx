import type { ActionLogDocument } from "@/services/rxdb/collections/actionLog";

import { For } from "solid-js";

import { useRxDBService } from "@/services/rxdb";
import { createSubscribeAllSignal } from "@/services/rxdb/subscribe";

export function ActionLogList() {
  const { collections } = useRxDBService();

  const actionLogs$ = createSubscribeAllSignal(() =>
    collections.actionLogs.find({
      selector: { beginAt: { $gt: 0 } },
      sort: [{ beginAt: "asc" }],
    })
  );

  const tentativeActionLogs$ = createSubscribeAllSignal(() => collections.actionLogs.find({ selector: { beginAt: 0 } }));

  return (
    <>
      <For each={actionLogs$()}>{(actionLog) => <ActionLog actionLog={actionLog} />}</For>
      <For each={tentativeActionLogs$()}>{(actionLog) => <ActionLog actionLog={actionLog} />}</For>
    </>
  );
}

function ActionLog(props: { actionLog: ActionLogDocument }) {
  return (
    <div>
      <span>{props.actionLog.beginAt}</span>
      <span>→</span>
      <span>{props.actionLog.endAt}</span>
      <span>：</span>
      <span>{props.actionLog.text}</span>
    </div>
  );
}
