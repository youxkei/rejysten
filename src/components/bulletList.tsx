import type { JSX } from "solid-js";

import { styles } from "@/styles.css";

export function BulletList(props: { bullet: JSX.Element; item: JSX.Element; child: JSX.Element; isSelected: boolean }) {
  return (
    <div class={styles.bulletList.container}>
      <div class={styles.bulletList.bullet}>{props.bullet}</div>
      <div classList={{ [styles.bulletList.item]: true, [styles.selected]: props.isSelected }}>{props.item}</div>
      <div class={styles.bulletList.child}>{props.child}</div>
    </div>
  );
}
