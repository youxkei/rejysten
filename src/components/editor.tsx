import { createEffect, onCleanup } from "solid-js";

import { useEventService } from "@/services/event";
import { createSignalWithLock, useLockService } from "@/services/lock";
import { useStoreService } from "@/services/store";
import { Input } from "@/solid/input";
import { styles } from "@/styles.css";

export function Editor(props: { text: string }) {
  let input!: HTMLInputElement;

  const lock = useLockService();
  const { store } = useStoreService();
  const { emitEvent } = useEventService();

  const cursorPosition$ = createSignalWithLock(lock, () => store.editor.cursorPosition, -1, true);
  const text$ = createSignalWithLock(lock, () => props.text, "", true);

  createEffect(() => {
    input.focus();

    const pos = cursorPosition$();

    if (pos === -1) {
      const len = props.text.length;
      input.setSelectionRange(len, len);
    } else {
      input.setSelectionRange(pos, pos);
    }
  });

  let active = true;
  onCleanup(() => {
    active = false;
  });

  const onBlur = () => {
    if (active) {
      emitEvent({ pane: store.currentPane, mode: "insert", type: "leaveInsertMode" });
    }
  };

  return (
    <Input
      class={styles.editor}
      ref={input}
      onInput={(event) => emitEvent({ pane: store.currentPane, mode: "insert", type: "changeEditorText", newText: event.target.value })}
      onBlur={onBlur}
      value={text$()}
    />
  );
}
