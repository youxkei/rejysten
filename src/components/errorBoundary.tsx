import type { JSX } from "solid-js";

import { ErrorBoundary as SolidErrorBoundary } from "solid-js";

async function resetDB() {
  for (const db of await window.indexedDB.databases()) {
    window.indexedDB.deleteDatabase(db.name!);
  }
}

export function ErrorBoundary(props: { children: JSX.Element }) {
  return (
    <SolidErrorBoundary
      fallback={(err: { stack: string }, reset) => (
        <>
          <p>Something went wrong.</p>
          <pre>{err.stack}</pre>
          <button onClick={reset}>reset</button>
          <button onClick={resetDB}>reset DB</button>
        </>
      )}
    >
      {props.children}
    </SolidErrorBoundary>
  );
}
