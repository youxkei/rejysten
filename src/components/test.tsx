import type { JSXElement } from "solid-js";

import { Suspense } from "solid-js";

import { RxDBServiceProvider } from "@/services/rxdb";
import { createDatabaseCreatorForTest } from "@/services/rxdb/test";

export function TestWithRxDBService(props: { tid: string; children: JSXElement }) {
  return (
    <Suspense fallback={props.tid}>
      <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>
        {props.children}
      </RxDBServiceProvider>
    </Suspense>
  );
}

export { createCollectionsForTest } from "@/services/rxdb/test";
