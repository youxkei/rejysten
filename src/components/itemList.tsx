import { Show, For, createMemo, createEffect } from "solid-js";

import { useCollections, useSubscribe, useSubscribeAll } from "@/rxdb";
import { BulletList } from "@/components/bulletList";

export function ItemList(props: { id: string }) {
  return (
    <BulletList
      bullet={"â€¢"}
      item={<ItemListItem id={props.id} />}
      child={<ItemListChildren parentId={props.id} />}
    />
  );
}

export function ItemListItem(props: { id: string }) {
  const collections = useCollections();
  const listItem = useSubscribe(() =>
    collections()?.listItems.findOne(props.id)
  );

  return (
    <Show when={listItem()}>
      <span>{listItem()!.text}</span>
    </Show>
  );
}

export function ItemListChildren(props: { parentId: string }) {
  const collections = useCollections();
  const children = useSubscribeAll(() =>
    collections()?.listItems.find({ selector: { parentId: props.parentId } })
  );

  const sortedChildren = createMemo(
    () => {
      type ListItem = ReturnType<typeof children>[0];

      const unsortedChildren = children();
      if (unsortedChildren.length === 0) {
        return { inconsistent: false, children: unsortedChildren };
      }

      const sortedChildren = [] as ListItem[];
      const childMap = new Map<string, ListItem>();
      let currentChildId: string | null = null;

      for (const child of unsortedChildren) {
        if (child.prevId === "") {
          currentChildId = child.id;
        }
        childMap.set(child.id, child);
      }

      if (currentChildId === null) {
        return { inconsistent: true, children: [] };
      }

      while (currentChildId !== "") {
        const currentChild = childMap.get(currentChildId);

        if (currentChild === undefined) {
          return { inconsistent: true, children: [] };
        }

        sortedChildren.push(currentChild);
        currentChildId = currentChild.nextId;
      }

      if (unsortedChildren.length != sortedChildren.length) {
        return { inconsistent: true, children: [] };
      }

      return { inconsistent: false, children: sortedChildren };
    },
    { inconsistent: false, children: [] },
    { equals: (_, next) => next.inconsistent }
  );

  return (
    <>
      <For each={sortedChildren().children}>
        {(child) => <ItemList id={child.id} />}
      </For>
    </>
  );
}

import {
  render,
  waitForElementToBeRemoved,
  queryByText,
  findByText,
} from "solid-testing-library";

import { TestWithRxDB, createCollections } from "@/rxdb/test";

if (import.meta.vitest) {
  describe.each([
    {
      name: "swapped",
      listItems: [
        {
          id: "001",
          text: "root",
          prevId: "",
          nextId: "",
          parentId: "",
        },
        {
          id: "002",
          text: "foo",
          prevId: "003",
          nextId: "",
          parentId: "001",
        },
        {
          id: "003",
          text: "bar",
          prevId: "",
          nextId: "002",
          parentId: "001",
        },
      ],
    },
    {
      name: "nested",
      listItems: [
        {
          id: "001",
          text: "root",
          prevId: "",
          nextId: "",
          parentId: "",
        },
        {
          id: "002",
          text: "foo",
          prevId: "",
          nextId: "003",
          parentId: "001",
        },
        {
          id: "0021",
          text: "foofoo",
          prevId: "",
          nextId: "",
          parentId: "002",
        },
        {
          id: "003",
          text: "bar",
          prevId: "002",
          nextId: "",
          parentId: "001",
        },
      ],
    },
  ])("$name", ({ listItems }) => {
    test("text changes", async (ctx) => {
      const tid = ctx.meta.id;
      let collections = await createCollections(tid);
      await collections.listItems.bulkUpsert(listItems);

      const { container, unmount } = render(() => (
        <TestWithRxDB tid={tid}>
          <ItemList id="001" />
        </TestWithRxDB>
      ));

      await waitForElementToBeRemoved(() => queryByText(container, tid));
      ctx.expect(container).toMatchSnapshot();

      await (await collections.listItems.findOne("001").exec())!.atomicPatch({
        text: "changed root",
      });

      await findByText(container, "changed root");
      ctx.expect(container).toMatchSnapshot();

      await (await collections.listItems.findOne("002").exec())!.atomicPatch({
        text: "changed foo",
      });

      await findByText(container, "changed foo");
      ctx.expect(container).toMatchSnapshot();

      unmount();
    });
  });

  test("add child", async (ctx) => {
    const tid = ctx.meta.id;
    let collections = await createCollections(tid);
    await collections.listItems.bulkUpsert([
      {
        id: "001",
        text: "root",
        prevId: "",
        nextId: "",
        parentId: "",
      },
      {
        id: "002",
        text: "foo",
        prevId: "",
        nextId: "003",
        parentId: "001",
      },
      {
        id: "003",
        text: "bar",
        prevId: "002",
        nextId: "",
        parentId: "001",
      },
    ]);

    const { container, unmount } = render(() => (
      <TestWithRxDB tid={tid}>
        <ItemList id="001" />
      </TestWithRxDB>
    ));

    await waitForElementToBeRemoved(() => queryByText(container, tid));
    ctx.expect(container).toMatchSnapshot();

    await collections.listItems.bulkUpsert([
      {
        id: "003",
        text: "bar",
        prevId: "002",
        nextId: "004",
        parentId: "001",
      },
      {
        id: "004",
        text: "baz",
        prevId: "003",
        nextId: "",
        parentId: "001",
      },
    ]);

    await findByText(container, "baz");
    ctx.expect(container).toMatchSnapshot();

    unmount();
  });
}
