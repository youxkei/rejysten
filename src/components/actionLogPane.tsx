import userEvent from "@testing-library/user-event";
import { Ulid } from "id128";
import { Show, onMount } from "solid-js";

import { ItemListChildren } from "@/components/itemList";
import { createSignalWithLock, runWithLock, useLockService, waitLockRelease } from "@/services/lock";
import { useRxDBService } from "@/services/rxdb";
import { getBottomItem } from "@/services/rxdb/collections/listItem";
import { createSubscribeSignal } from "@/services/rxdb/subscribe";
import { useStoreService } from "@/services/store";
import { renderWithServicesForTest } from "@/services/test";
import { shortenClassName } from "@/test";

export function ActionLogPane() {
  const rxdb = useRxDBService();
  const lock = useLockService();
  const { store, updateStore } = useStoreService();

  onMount(() => {
    if (store.actionLogPane.currentListItemId !== "") return;

    void runWithLock(lock, async () => {
      const bottomItem = await getBottomItem(rxdb, store.actionLogPane.currentActionLogId);

      let id = "";
      if (bottomItem) {
        id = bottomItem.id;
      } else {
        const now = Date.now();
        id = Ulid.generate({ time: now }).toCanonical();

        await rxdb.collections.listItems.insert({
          id,
          text: "",
          nextId: "",
          prevId: "",
          parentId: store.actionLogPane.currentActionLogId,
          updatedAt: Date.now(),
        });
      }

      await updateStore((store) => {
        store.actionLogPane.currentListItemId = id;
      });
    });
  });

  const actionLog$ = createSignalWithLock(
    lock,
    createSubscribeSignal(() => rxdb.collections.actionLogs.findOne(store.actionLogPane.currentActionLogId)),
    null
  );

  return (
    <Show when={actionLog$()}>
      {(actionLog$) => (
        <>
          {actionLog$().text}
          <ItemListChildren parentId={store.actionLogPane.currentActionLogId} selectedId={store.actionLogPane.currentListItemId} />
        </>
      )}
    </Show>
  );
}

if (import.meta.vitest) {
  describe("display", () => {
    test("normal mode", async (test) => {
      const { container, unmount } = await renderWithServicesForTest(
        test.meta.id,
        (props) => (
          <>
            <ActionLogPane />
            {props.children}
          </>
        ),
        async ({ rxdb: { collections }, store: { updateStore } }) => {
          await collections.actionLogs.insert({ id: "log1", text: "log1", startAt: 0, endAt: 0, updatedAt: 0 });
          await collections.listItems.bulkInsert([
            { id: "item1", text: "item1", parentId: "log1", prevId: "", nextId: "item2", updatedAt: 0 },
            /**/ { id: "item1_1", text: "item1_1", parentId: "item1", prevId: "", nextId: "item1_2", updatedAt: 0 },
            /**/ { id: "item1_2", text: "item1_2", parentId: "item1", prevId: "item1_1", nextId: "", updatedAt: 0 },
            { id: "item2", text: "item2", parentId: "log1", prevId: "item1", nextId: "", updatedAt: 0 },
            /**/ { id: "item2_1", text: "item2_1", parentId: "item2", prevId: "", nextId: "", updatedAt: 0 },
            /*     */ { id: "item2_1_1", text: "item2_1_1", parentId: "item2_1", prevId: "", nextId: "", updatedAt: 0 },
          ]);
          await updateStore((store) => {
            store.currentPane = "actionLog";
            store.actionLogPane.currentActionLogId = "log1";
            store.actionLogPane.currentListItemId = "item1_2";
          });
        }
      );

      test.expect(shortenClassName(container)).toMatchSnapshot();

      unmount();
    });

    test("insert mode", async (test) => {
      const { container, unmount, findByDisplayValue } = await renderWithServicesForTest(
        test.meta.id,
        (props) => (
          <>
            <ActionLogPane />
            {props.children}
          </>
        ),
        async ({ rxdb: { collections }, store: { updateStore } }) => {
          await collections.actionLogs.insert({ id: "log1", text: "log1", startAt: 0, endAt: 0, updatedAt: 0 });
          await collections.listItems.bulkInsert([
            { id: "item1", text: "item1", parentId: "log1", prevId: "", nextId: "item2", updatedAt: 0 },
            /**/ { id: "item1_1", text: "item1_1", parentId: "item1", prevId: "", nextId: "item1_2", updatedAt: 0 },
            /**/ { id: "item1_2", text: "item1_2", parentId: "item1", prevId: "item1_1", nextId: "", updatedAt: 0 },
            { id: "item2", text: "item2", parentId: "log1", prevId: "item1", nextId: "", updatedAt: 0 },
            /**/ { id: "item2_1", text: "item2_1", parentId: "item2", prevId: "", nextId: "", updatedAt: 0 },
            /*     */ { id: "item2_1_1", text: "item2_1_1", parentId: "item2_1", prevId: "", nextId: "", updatedAt: 0 },
          ]);
          await updateStore((store) => {
            store.currentPane = "actionLog";
            store.mode = "insert";
            store.editor.text = "item1_2";
            store.editor.cursorPosition = 3; // ite|m1_2
            store.actionLogPane.currentActionLogId = "log1";
            store.actionLogPane.currentListItemId = "item1_2";
          });
        }
      );

      test.expect(shortenClassName(container)).toMatchSnapshot();

      const input = await findByDisplayValue<HTMLInputElement>("item1_2");
      test.expect(input.selectionStart).toBe(3);
      test.expect(input.selectionEnd).toBe(3);

      unmount();
    });
  });

  describe("PC operations", () => {
    describe("keyboard operations", () => {
      describe.each([
        { name: "press i to enter insert mode", key: "i", text: "text", wantCursorPosition: 0 },
        { name: "press a to enter insert mode", key: "a", text: "text", wantCursorPosition: 4 },
      ])("$name", ({ key, text, wantCursorPosition }) => {
        test("assert", async (test) => {
          const user = userEvent.setup();
          const { container, unmount, findByDisplayValue } = await renderWithServicesForTest(
            test.meta.id,
            (props) => (
              <>
                <ActionLogPane />
                {props.children}
              </>
            ),
            async ({ rxdb: { collections }, store: { updateStore } }) => {
              await collections.actionLogs.insert({ id: "root", text: "root", startAt: 0, endAt: 0, updatedAt: 0 });
              await collections.listItems.insert({ id: "1", text, parentId: "root", prevId: "", nextId: "", updatedAt: 0 });
              await updateStore((store) => {
                store.currentPane = "actionLog";
                store.actionLogPane.currentActionLogId = "root";
                store.actionLogPane.currentListItemId = "1";
              });
            }
          );

          await user.keyboard(key);

          const input = await findByDisplayValue<HTMLInputElement>("text");

          test.expect(shortenClassName(container)).toMatchSnapshot("after press " + key);
          test.expect(input.selectionStart).toBe(wantCursorPosition);
          test.expect(input.selectionEnd).toBe(wantCursorPosition);

          unmount();
        });
      });

      describe.each([
        { name: "press o to add item below", key: "o" },
        { name: "press O to add item above", key: "{Shift>}o{/Shift}" },
      ])("$name", ({ key }) => {
        test("assert", async (test) => {
          const user = userEvent.setup();
          const { container, unmount, findByRole } = await renderWithServicesForTest(
            test.meta.id,
            (props) => (
              <>
                <ActionLogPane />
                {props.children}
              </>
            ),
            async ({ rxdb: { collections }, store: { updateStore } }) => {
              await collections.actionLogs.insert({ id: "root", text: "root", startAt: 0, endAt: 0, updatedAt: 0 });
              await collections.listItems.insert({ id: "1", text: "text", parentId: "root", prevId: "", nextId: "", updatedAt: 0 });
              await updateStore((store) => {
                store.currentPane = "actionLog";
                store.actionLogPane.currentActionLogId = "root";
                store.actionLogPane.currentListItemId = "1";
              });
            }
          );

          await user.keyboard(key);

          const input = await findByRole<HTMLInputElement>("textbox");

          test.expect(shortenClassName(container)).toMatchSnapshot("after press " + key);
          test.expect(input.selectionStart).toBe(0);
          test.expect(input.selectionEnd).toBe(0);

          unmount();
        });
      });

      describe.each([
        { name: "press Tab in normal mode to indent item", key: "{Tab}" },
        { name: "press Shift+Tab in normal mode to dedent item", key: "{Shift>}{Tab}{/Shift}" },
      ])("$name", ({ key }) => {
        test("assert", async (test) => {
          const user = userEvent.setup();
          const { container, unmount, lock } = await renderWithServicesForTest(
            test.meta.id,
            (props) => (
              <>
                <ActionLogPane />
                {props.children}
              </>
            ),
            async ({ rxdb: { collections }, store: { updateStore } }) => {
              await collections.actionLogs.insert({ id: "log1", text: "log1", startAt: 1000, endAt: 0, updatedAt: 0 });
              await collections.listItems.bulkInsert([
                { id: "item1", text: "item1", parentId: "log1", prevId: "", nextId: "", updatedAt: 0 },
                { id: "item2", text: "item2", parentId: "item1", prevId: "", nextId: "item3", updatedAt: 0 },
                { id: "item3", text: "item3", parentId: "item1", prevId: "item2", nextId: "", updatedAt: 0 },
              ]);

              await updateStore((store) => {
                store.currentPane = "actionLog";
                store.actionLogPane.currentActionLogId = "log1";
                store.actionLogPane.currentListItemId = "item3";
              });
            }
          );

          await user.keyboard(key);
          await waitLockRelease(lock);

          test.expect(shortenClassName(container)).toMatchSnapshot("after press " + key);

          unmount();
        });
      });

      describe.each([
        { name: "press Tab in insert mode to indent item", key: "{Tab}" },
        { name: "press Shift+Tab in insert mode to dedent item", key: "{Shift>}{Tab}{/Shift}" },
      ])("$name", ({ key }) => {
        test("assert", async (test) => {
          const cursorPosition = Math.floor(Math.random() * ("item3".length + 1));

          const user = userEvent.setup();
          const { container, unmount, lock, findByDisplayValue } = await renderWithServicesForTest(
            test.meta.id,
            (props) => (
              <>
                <ActionLogPane />
                {props.children}
              </>
            ),
            async ({ rxdb: { collections }, store: { updateStore } }) => {
              await collections.actionLogs.insert({ id: "log1", text: "log1", startAt: 1000, endAt: 0, updatedAt: 0 });
              await collections.listItems.bulkInsert([
                { id: "item1", text: "item1", parentId: "log1", prevId: "", nextId: "", updatedAt: 0 },
                { id: "item2", text: "item2", parentId: "item1", prevId: "", nextId: "item3", updatedAt: 0 },
                { id: "item3", text: "item3", parentId: "item1", prevId: "item2", nextId: "", updatedAt: 0 },
              ]);

              await updateStore((store) => {
                store.currentPane = "actionLog";
                store.mode = "insert";
                store.editor.text = "item3";
                store.editor.cursorPosition = cursorPosition;
                store.actionLogPane.currentActionLogId = "log1";
                store.actionLogPane.currentListItemId = "item3";
              });
            }
          );

          await user.keyboard(key);
          await waitLockRelease(lock);

          const input = await findByDisplayValue<HTMLInputElement>("item3");

          test.expect(shortenClassName(container)).toMatchSnapshot("after press " + key);
          test.expect(input.selectionStart).toBe(cursorPosition);
          test.expect(input.selectionEnd).toBe(cursorPosition);

          unmount();
        });
      });

      test("press Backspace to remove a character", async (test) => {
        const user = userEvent.setup();
        const { container, unmount, lock, findByDisplayValue } = await renderWithServicesForTest(
          test.meta.id,
          (props) => (
            <>
              <ActionLogPane />
              {props.children}
            </>
          ),
          async ({ rxdb: { collections }, store: { updateStore } }) => {
            await collections.actionLogs.insert({ id: "log1", text: "log1", startAt: 1000, endAt: 0, updatedAt: 0 });
            await collections.listItems.bulkInsert([
              { id: "item1", text: "item1", parentId: "log1", prevId: "", nextId: "item2", updatedAt: 0 },
              { id: "item2", text: "item2", parentId: "log1", prevId: "item1", nextId: "", updatedAt: 0 },
            ]);

            await updateStore((store) => {
              store.currentPane = "actionLog";
              store.mode = "insert";
              store.editor.text = "item2";
              store.editor.cursorPosition = 3; // ite|m2
              store.actionLogPane.currentActionLogId = "log1";
              store.actionLogPane.currentListItemId = "item2";
            });
          }
        );

        await user.keyboard("{Backspace}");

        const input = await findByDisplayValue<HTMLInputElement>("itm2");

        test.expect(shortenClassName(container)).toMatchSnapshot("after press Backspace");
        test.expect(input.selectionStart).toBe(2);
        test.expect(input.selectionEnd).toBe(2);

        unmount();
      });

      describe("press Backspace to remove item", () => {
        test.skip("cursor is not on the left edge: item is not removed", () => {
          // TODO
        });

        test.skip("cursor is on the left edge, has children: item is not removed", () => {
          // TODO
        });

        test.skip("cursor is on the left edge, no children, no above item, has below item: item is not removed", () => {
          // TODO
        });

        test.skip("cursor is on the left edge, no children, has above item: item is removed and move to above item", () => {
          // TODO
        });

        test.skip("text is empty, no parent, no children, no above item, no below item: item is removed and move to ActionLogListPane", () => {
          // TODO
        });
      });

      describe("press Delete to remove below item", () => {
        test.skip("cursor is not on the right edge: below item is not removed", () => {
          // TODO
        });

        test.skip("cursor is on the right edge, no below item: nothing happened", () => {
          // TODO
        });

        test.skip("cursor is on the right edge, below item has children: below item is not removed", () => {
          // TODO
        });

        test.skip("cursor is on the right edge, below item has no children: below item is removed text is merged", () => {
          // TODO
        });
      });

      describe("press Enter to add item", () => {
        test.skip("cursor is on the right edge: empty item is added and move to added item", () => {
          // TODO
        });

        test.skip("cursor is in the middle of text: non-empty item is added and move to added item", () => {
          // TODO
        });
      });
    });
  });
}
