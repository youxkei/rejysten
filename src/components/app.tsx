import type { ChangeEvent } from "react";

import { useSelector, useDispatch } from "@/store";
import { app } from "@/slices/app";
import { useRxSync } from "@/rxdb";
import { RxdbSyncConfig } from "@/components/rxdbSyncConfig";
import { ItemList } from "@/components/itemList";

export function App() {
  useRxSync();

  const id = useSelector((state) => state.app.id);
  const dispatch = useDispatch();

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch(app.actions.updateId({ id: event.target.value }));
  };

  return (
    <>
      {id === "" ? null : <ItemList id={id} />}
      <input value={id} onChange={onChange} />
      <RxdbSyncConfig />
    </>
  );
}
