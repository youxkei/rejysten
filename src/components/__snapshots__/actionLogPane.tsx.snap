// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PC operations > keyboard operations > 'press O to add item above' > assert > after press {Shift>}o{/Shift} 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <input
        class="editor"
      />
    </div>
    <div
      class="bulletList_child"
    />
  </div>
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press O to add item above' > assert > initial 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press Shift+Tab in normal mode to dedent item' > assert > after press {Shift>}{Tab}{/Shift} 1`] = `
<div>
  log1
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item"
    >
      <span>
        item1
      </span>
    </div>
    <div
      class="bulletList_child"
    >
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item"
        >
          <span>
            item2
          </span>
        </div>
        <div
          class="bulletList_child"
        />
      </div>
    </div>
  </div>
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        item3
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press Shift+Tab in normal mode to dedent item' > assert > initial 1`] = `
<div>
  log1
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item"
    >
      <span>
        item1
      </span>
    </div>
    <div
      class="bulletList_child"
    >
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item"
        >
          <span>
            item2
          </span>
        </div>
        <div
          class="bulletList_child"
        />
      </div>
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item selected"
        >
          <span>
            item3
          </span>
        </div>
        <div
          class="bulletList_child"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press Tab in normal mode to indent item' > assert > after press {Tab} 1`] = `
<div>
  log1
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item"
    >
      <span>
        item1
      </span>
    </div>
    <div
      class="bulletList_child"
    >
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item"
        >
          <span>
            item2
          </span>
        </div>
        <div
          class="bulletList_child"
        >
          <div
            class="bulletList_container"
          >
            <div
              class="bulletList_bullet"
            >
              •
            </div>
            <div
              class="bulletList_item selected"
            >
              <span>
                item3
              </span>
            </div>
            <div
              class="bulletList_child"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press Tab in normal mode to indent item' > assert > initial 1`] = `
<div>
  log1
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item"
    >
      <span>
        item1
      </span>
    </div>
    <div
      class="bulletList_child"
    >
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item"
        >
          <span>
            item2
          </span>
        </div>
        <div
          class="bulletList_child"
        />
      </div>
      <div
        class="bulletList_container"
      >
        <div
          class="bulletList_bullet"
        >
          •
        </div>
        <div
          class="bulletList_item selected"
        >
          <span>
            item3
          </span>
        </div>
        <div
          class="bulletList_child"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press a to enter insert mode' > assert > after press a 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <input
        class="editor"
      />
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press a to enter insert mode' > assert > initial 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press i to enter insert mode' > assert > after press i 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <input
        class="editor"
      />
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press i to enter insert mode' > assert > initial 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press o to add item below' > assert > after press o 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <input
        class="editor"
      />
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;

exports[`PC operations > keyboard operations > 'press o to add item below' > assert > initial 1`] = `
<div>
  root
  <div
    class="bulletList_container"
  >
    <div
      class="bulletList_bullet"
    >
      •
    </div>
    <div
      class="bulletList_item selected"
    >
      <span>
        text
      </span>
    </div>
    <div
      class="bulletList_child"
    />
  </div>
</div>
`;
